File: 2018-minimum-space-wasted-from-packaging/minimum-space-wasted-from-packaging.cs
public class Solution {
    public int MinWastedSpace(int[] packages, int[][] boxes) {
        const int MOD = 1000000007;

        Array.Sort(packages);

        int n = packages.Length;

        long[] prefix = new long[n + 1];

        for (int i = 0; i < n; i++)
            prefix[i + 1] = prefix[i] + packages[i];

        long ans = long.MaxValue;

        foreach (var supplier in boxes) {
            Array.Sort(supplier);

            if (supplier[supplier.Length - 1] < packages[n - 1])
                continue;

            long waste = 0;
            int idx = 0;

            foreach (int b in supplier) {
                int next = UpperBound(packages, idx, b);

                waste += (long)(next - idx) * b - (prefix[next] - prefix[idx]);

                idx = next;
                if (idx == n) break;
            }

            ans = Math.Min(ans, waste);
        }

        return ans == long.MaxValue ? -1 : (int)(ans % MOD);
    }

    private int UpperBound(int[] arr, int start, int target) {
        int l = start, r = arr.Length;

        while (l < r) {
            int mid = (l + r) / 2;

            if (arr[mid] <= target)
                l = mid + 1;
            else
                r = mid;
        }

        return l;
    }
}

File: 2018-minimum-space-wasted-from-packaging/minimum-space-wasted-from-packaging.scala
object Solution {
    def minWastedSpace(packages: Array[Int], boxes: Array[Array[Int]]): Int = {
        val MOD = 1000000007L

        scala.util.Sorting.quickSort(packages)

        val n = packages.length

        val prefix = new Array[Long](n + 1)
        for (i <- 0 until n)
            prefix(i + 1) = prefix(i) + packages(i)

        var ans = Long.MaxValue

        for (supplier <- boxes) {
            scala.util.Sorting.quickSort(supplier)

            if (supplier.last >= packages.last) {
                var waste = 0L
                var idx = 0

                for (b <- supplier if idx < n) {
                    val next = upperBound(packages, idx, b)

                    waste += (next - idx).toLong * b - (prefix(next) - prefix(idx))

                    idx = next
                }

                ans = math.min(ans, waste)
            }
        }

        if (ans == Long.MaxValue) -1
        else (ans % MOD).toInt
    }

    def upperBound(arr: Array[Int], start: Int, target: Int): Int = {
        var l = start
        var r = arr.length

        while (l < r) {
            val mid = (l + r) / 2
            if (arr(mid) <= target)
                l = mid + 1
            else
                r = mid
        }

        l
    }
}